cpp/src/arrow/compute/cast.cc

```diff
@@ -300,11 +560,14 @@ class CastKernel : public UnaryKernel {
   CastOptions options_;
   CastFunction func_;
   bool is_zero_copy_;
+  bool can_pre_allocate_values_;
 };
 
 #define CAST_CASE(InType, OutType)                                                  \
   case OutType::type_id:                                                            \
     is_zero_copy = is_zero_copy_cast<OutType, InType>::value;                       \
+    can_pre_allocate_values =                                                       \
+        !(!is_binary_like(InType::type_id) && is_binary_like(OutType::type_id));    \
```

Comment:
Aside: we'll want to refactor this bit so be all based on constexprs and written in a way that the optimizer will be able to trim dead code in release builds (i.e. if it is not zero copy, there is no need to assign a value to `is_zero_copy`)